File: FBDeviceControl/Management/FBDeviceStorage.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDeviceStorage.h"

@interface FBDeviceStorage ()

@property (nonatomic, strong, readonly) id<FBControlCoreLogger> logger;
@property (nonatomic, strong, readonly) NSMutableDictionary<NSString *, id> *attachedDevices;
@property (nonatomic, strong, readonly) NSMapTable<NSString *, id> *referencedDevices;

@end

@implementation FBDeviceStorage

- (instancetype)initWithLogger:(id<FBControlCoreLogger>)logger
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _logger = logger;
  _attachedDevices = [NSMutableDictionary dictionary];
  _referencedDevices = [NSMapTable mapTableWithKeyOptions:NSPointerFunctionsCopyIn valueOptions:NSPointerFunctionsWeakMemory];

  return self;
}

#pragma mark Public Methods

- (void)deviceAttached:(id)device forKey:(NSString *)key
{
  // Set both the strong-memory and the weak-memory device.
  // If it already exists this is fine, otherwise it will ensure that this mapping is preserved.
  // Any removed devies will be removed from attachedDevices on disconnect so that abandoned device references are cleaned up.
  id attached = self.attachedDevices[key];
  id referenced = [self.referencedDevices objectForKey:key];
  if (attached && referenced) {
    [self.logger logFormat:@"%@ is an attached device update", device];
  } else if (referenced) {
    [self.logger logFormat:@"%@ is referenced and now attached again", device];
  } else {
    [self.logger logFormat:@"%@ appeared for the first time", device];
  }
  self.attachedDevices[key] = device;
  [self.referencedDevices setObject:device forKey:key];
}

- (void)deviceDetachedForKey:(NSString *)key
{
  // Remove only from the list of attached devices.
  // If the device instance is not referenced elsewhere it will be removed from the referencedDevices dictionary.
  // This is because the values in that dictionary are weakly referenced.
  [self.attachedDevices removeObjectForKey:key];
}

- (nullable id)deviceForKey:(NSString *)key
{
  return self.attachedDevices[key] ?: [self.referencedDevices objectForKey:key];
}

#pragma mark Public

- (NSDictionary<NSString *, id> *)attached
{
  return [self.attachedDevices copy];
}

- (NSDictionary<NSString *, id> *)referenced
{
  return self.referencedDevices.dictionaryRepresentation;
}

@end

File: FBDeviceControl/Management/FBDeviceStorage.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import FBControlCore
import Foundation

/// Holds both the currently-attached devices and weak references to every device this storage has
/// ever vended, so a consumer holding a device across a disconnect gets the same instance back on
/// re-attach.
public final class FBDeviceStorage<T: AnyObject> {

  public var attached: [String: T] {
    attachedDevices
  }

  public var referenced: [String: T] {
    var result: [String: T] = [:]
    let enumerator = referencedDevices.keyEnumerator()
    while let key = enumerator.nextObject() as? NSString {
      if let value = referencedDevices.object(forKey: key) as? T {
        result[key as String] = value
      }
    }
    return result
  }

  private let logger: any FBControlCoreLogger
  private var attachedDevices: [String: T]
  private var referencedDevices: NSMapTable<NSString, AnyObject>

  public init(logger: any FBControlCoreLogger) {
    self.logger = logger
    self.attachedDevices = [:]
    self.referencedDevices = NSMapTable(keyOptions: .copyIn, valueOptions: .weakMemory)
  }

  public func deviceAttached(_ device: T, forKey key: String) {
    let attached = attachedDevices[key]
    let referenced = referencedDevices.object(forKey: key as NSString)
    if attached != nil && referenced != nil {
      logger.log("\(device) is an attached device update")
    } else if referenced != nil {
      logger.log("\(device) is referenced and now attached again")
    } else {
      logger.log("\(device) appeared for the first time")
    }
    attachedDevices[key] = device
    referencedDevices.setObject(device, forKey: key as NSString)
  }

  public func deviceDetached(forKey key: String) {
    attachedDevices.removeValue(forKey: key)
  }

  public func device(forKey key: String) -> T? {
    attachedDevices[key] ?? referencedDevices.object(forKey: key as NSString) as? T
  }
}
